### PR TITLE
Mcts core dev

### PR DESCRIPTION
**Pull request by Hoeiriis**
Implemented a general interface for MCTS: MCTSInterface as well as implementing the basic UCT algorithm

### CODE REVIEW

File: src/Core/mcts/MCTSInterface.cpp [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+        // TreePolicy runs to find an unexpanded node to expand
+        auto expandedNode = m_tree_policy(m_root);
+        // From the expanded node, a simulation runs that returns a score
+        Reward simulation_score = m_default_policy(expandedNode->state);
+        // The score is backpropagated up through the search tree
+        m_backpropagation(expandedNode, simulation_score);
+    }
+
+    return m_best_child(m_root, 0);
+}
```
Comment (Operdies) [inline src/Core/mcts/MCTSInterface.cpp]: Should C++ interfaces provide implementations?

File: src/Core/mcts/MCTSBase.cpp [hunk 2]
```diff
@@ -1,6 +1,33 @@
...
+
+std::shared_ptr<SearchNode> MCTSBase::search(int n_searches) {
+
+    for (int i = 0; i < n_searches; i++) {
+        // TreePolicy runs to find an unexpanded node to expand
+        auto node_to_expand = m_tree_policy(m_root);
+        // From the unexpanded node, a simulation runs that returns a score
+        Reward simulation_score = m_default_policy(node_to_expand->state);
+        // The score is backpropagated up through the search tree
+        m_backpropagation(node_to_expand, simulation_score);
```
Comment (Operdies) [inline src/Core/mcts/MCTSBase.cpp]: No return in this function

File: src/Core/mcts/MCTSInterface.cpp [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+    }
+
+    return m_root->state;
+}
+
+std::shared_ptr<SearchNode> MCTSInterface::m_search(int n_searches) {
+
+    for (int i = 0; i < n_searches; i++) {
+        // TreePolicy runs to find an unexpanded node to expand
+        auto expandedNode = m_tree_policy(m_root);
```
Comment (Operdies) [inline src/Core/mcts/MCTSInterface.cpp]: can't figure out where the tree policy is defined

File: src/Core/mcts/UCT.cpp [hunk 4]
```diff
@@ -1,6 +1,57 @@
...
+UCT::UCT(EnvironmentBase &environment) : MCTSInterface(environment), generator(std::mt19937(time(nullptr))){};
 
-void UCT::run(int n_searches) {}
+std::shared_ptr<SearchNode> UCT::m_tree_policy(std::shared_ptr<SearchNode> node) { return m_tpolicy.treePolicy(node); };
+
+Reward UCT::m_default_policy(State &state) { return m_defaultPolicy.defaultPolicy(state); };
+
+std::shared_ptr<SearchNode> UCT::m_best_child(std::shared_ptr<SearchNode> node, double c) {
+    int best_child = 0;
+    double best_score_so_far = -1;
```
Comment (Operdies) [inline src/Core/mcts/UCT.cpp]: Depending on the domain, `-1` may not be a suitable score.
Use `std::numeric_limits<double>::lowest()`

Comment (Operdies) [reply src/Core/mcts/UCT.cpp]: Should be in `#include <limits>`, may need to google

File: src/Core/mcts/UCT.cpp [hunk 5]
```diff
@@ -1,6 +1,57 @@
...
+
+    for (int i = 0; i < node->child_nodes.size(); i++) {
+        auto child = node->child_nodes.at(i);
+        double score = (child->score / (child->visits + 0.000000001)) +
+                       c * std::sqrt((std::log(node->visits) / (child->visits + 0.000000001)));
```
Comment (Operdies) [inline src/Core/mcts/UCT.cpp]: Consider using `DBL_MIN` instead of 0.0000000001

File: src/Core/mcts/UCT.cpp [hunk 6]
```diff
@@ -1,6 +1,57 @@
...
+
+        if (score > best_score_so_far) {
+            best_score_so_far = score;
+            best_child = i;
+        } else if (score == best_score_so_far) {
```
Comment (Operdies) [inline src/Core/mcts/UCT.cpp]: Is this necessary to account for? There's a lot of doubles

File: src/Core/mcts/UCT.cpp [hunk 7]
```diff
@@ -1,6 +1,57 @@
...
+    if (node->unvisited_child_states.size() > 1) {
+        // Get unvisited state randomly
+        std::uniform_int_distribution<int> uniformIntDistribution(0, node->unvisited_child_states.size() - 1);
+        i_random = uniformIntDistribution(generator);
+    }
+
+    State expanded_state = node->unvisited_child_states.at(i_random);
+
+    // Create node from unvisited state
+    auto is_terminal = m_environment.GetValidChildStates(expanded_state).empty();
```
Comment (Operdies) [inline src/Core/mcts/UCT.cpp]: i think m_environment implements an is_terminal function

File: src/Core/mcts/include/UCT.h [hunk 8]
```diff
@@ -1,21 +1,51 @@
...
-{
-    private:
-        /* data */
-    public:
-        UCT(EnvironmentBase& environement, boost::function<int(State)> default_policy);
-        ~UCT();
+class UCT : MCTSInterface {
+  public:
+    explicit UCT(EnvironmentBase &environment);
+    ;
```
Comment (Operdies) [inline src/Core/mcts/include/UCT.h]: double semi colon?

File: src/Core/tree_policies/UCT_TreePolicy.cpp [hunk 9]
```diff
@@ -9,7 +10,8 @@ std::shared_ptr<SearchNode> UCT_TreePolicy::treePolicy(std::shared_ptr<SearchNod
             return m_expand(current_node);
         }
 
-        current_node = m_bestChild(current_node);
+        double cp = 1 / std::sqrt(2);
```
Comment (Operdies) [inline src/Core/tree_policies/UCT_TreePolicy.cpp]:
```suggestion
        constexpr sqrt2 = std::sqrt(2);
        double cp = 1 / sqrt2;
```
Slight performance increase

Comment (Operdies) [reply src/Core/tree_policies/UCT_TreePolicy.cpp]: constexpr is computed at compile time, and sqrt is a relatively expensive operation

File: src/Core/mcts/UCT.cpp (hunk 5 above)
Comment (Operdies) [reply src/Core/mcts/UCT.cpp]: `DBL_MIN` is in `values.h>`

Comment (Operdies) [reply src/Core/mcts/UCT.cpp]: On my machine:
`std::numeric_limits<double>::lowest() = -1.79769e+308`
`DBL_MIN = 2.22507e-308`

File: src/Core/mcts/MCTSBase.cpp (hunk 2 above)
Comment (Hoeiriis) [reply src/Core/mcts/MCTSBase.cpp]: It doesn't return anything, simply updates the ancestor nodes of the given node

File: src/Core/mcts/MCTSInterface.cpp (hunk 1 above)
Comment (Hoeiriis) [reply src/Core/mcts/MCTSInterface.cpp]: Hmm, good point

File: src/Core/mcts/UCT.cpp (hunk 7 above)
Comment (Hoeiriis) [reply src/Core/mcts/UCT.cpp]: It probably should, but currently it's evaluate_terminal_state, which may not necessarily be implemented in a way that can be used to check for terminality

File: src/Core/tree_policies/UCT_TreePolicy.cpp (hunk 9 above)
Comment (Hoeiriis) [reply src/Core/tree_policies/UCT_TreePolicy.cpp]: Coolio

File: src/Core/mcts/MCTSInterface.cpp (hunk 3 above)
Comment (Hoeiriis) [reply src/Core/mcts/MCTSInterface.cpp]: Virtual member functions defined in the header and implemented in deriving classes

**Pull request merged**